### ISSUE TITLE
How to deal with datasets which have both length measurements and abundances?

### ISSUE THREAD

**Issue by Daphnisd** [iobis/env-data#10]
The dataset has both abundance values, biomass values and length measurements for each specimen.
We can have one record occurrence record which can link to the abundance and the biomass. However how can you link the different length measurements?
I see 3 options:
1) link the parameter length X number of times to the same occurrence record. The format documentation would state that this refers to X different specimens
2) Record different occurrences for each length measurement; however there should be some link between the abundances and these "sub"occurrences
3) Don't include the individual length measurements; instead include derived parameters like average, median, min/max, SD,...

[For Example]( https://github.com/iobis/env-data/blob/master/datasets/ISSUE_%20HowToDealWithLengthsAndAbundances.xlsx?raw=true):

I think this issue can/should be handled together with the format for stomach contents. These are also related occurrences.
Issue also raised here: https://github.com/tdwg/dwc-qa/issues/76

**Comment by wardappeltans** [comment]
Dear Daphnis, 
We would opt for 1, same occurrence record for the different length measurements. 

Only if the dataset provides different organismIDs for each length measurement, you could opt for 2: different occurrences for each measurement (abundance, biomass, length), which are linked through the same eventID.  

We would _not_ go for option 3 (change the values to derived ones).

**Comment by AbbyBenson** [comment]
I had to do something like this for the [USGS GLSC Trawl](https://www1.usgs.gov/obis-usa/ipt/resource?r=usgs_glsc_rvcat_trawl). Sounds like I did it right? I have multiple lengths recorded for the same occurrenceID that relates to multiple specimens of the same species in a catch. So for instance:

**occurrenceID**                                           **individualCount**   **lifeStage**             **scientificName**
ca595e83-1fb1-11e7-9b31-319b2112c46a                   5                       Yearling                    Perca flavescens

**occurrenceID                                                              measurementValue          measurementUnit**
ca595e83-1fb1-11e7-9b31-319b2112c46a                                 71                           mm
ca595e83-1fb1-11e7-9b31-319b2112c46a                                 76                           mm
ca595e83-1fb1-11e7-9b31-319b2112c46a                                 67                           mm
ca595e83-1fb1-11e7-9b31-319b2112c46a                                 64                           mm
ca595e83-1fb1-11e7-9b31-319b2112c46a                                 12                             g

**Comment by Daphnisd** [comment]
Discussion with Mary about this. Proposed sollution: include a new collumn in the MoF "organismID" or similar. 
The occurrence would always refer to the abuncance level. Differrent measurements of different specimens would relate to the same occurrence id. Inclusion of a grouping paramer like "OrganismID" will allow for specifying which measurements belong to the same specimen. You can have for example.

**Occurrence:**
OccurrenceID = occ1, ScientificName= Gadus morhua, lifestage = adult, sex = female

**eMoF**
OccurrenceID = occ1,  measurementtype = abundance, measurementValue=  15 , measurementUnit = #/l

OccurrenceID = occ1,  organismid = org1, measurementtype = length, measurementValue=  15 , measurementUnit = cm
OccurrenceID = occ1,  organismid = org1, measurementtype = wifth, measurementValue=  3, measurementUnit = cm
OccurrenceID = occ1,  organismid = org1, measurementtype = weigth, measurementValue=  150 , measurementUnit = g 
OccurrenceID = occ1,  organismid = org1, measurementtype = individualcount, measurementValue=  3 

OccurrenceID = occ1,  organismid = org2, measurementtype = length, measurementValue=  11 , measurementUnit = cm
OccurrenceID = occ1,  organismid = org2, measurementtype = wifth, measurementValue=  2, measurementUnit = cm
OccurrenceID = occ1,  organismid = org2, measurementtype = weigth, measurementValue=  120 , measurementUnit = g 
OccurrenceID = occ1,  organismid = org2, measurementtype = individualcount, measurementValue=  5

**Comment by Daphnisd** [comment]
Another dataset dealing with this issue (from IMR):

> One question: ( I don’t think that it is relevant for the king crab, but I know that it is relevant for shrimps )
> I have a fish station with total catch of lets say 5000 shrimps. Of these 5000, I have length measurements of let say 100. How should I export this information? 
> 100 lines with individualCount=1 and one line with individualCount=4900?

Atm the answer is the following. It will work as there is only a single measurement per specimen.
1 occurrence record. 
1 eMoF record: Count = 5000
100 eMoF records: lenght = x

**Comment by pieterprovoost** [comment]
Here's a graphical summary of the above proposals. Proposal 1 seems fine in case of just one measurement. Is there any reason not to use proposal 2 in all cases? I'm still working on the database modelling and API design for measurements, but I expect that I'll be able to handle both proposals without too much overhead.

![abundance_biometric](https://user-images.githubusercontent.com/490041/49504713-912acd00-f87a-11e8-98c7-e9a1773c1716.png)

**Comment by Daphnisd** [comment]
I like 2. But not sure how it works for use case: https://github.com/iobis/env-data/issues/10#issuecomment-354739951? 

Also a similar dataset where there is a global abundance for all length(classes) and a narrower abundance one for each individual lenght(classes).

**Comment by rubenpp7** [comment]
Hi, 

The field organismID is in the Occurrence extension (and not in the eMoF extension) of IPT. How can we link the measurements to individual specimens without using different occurrenceIDs?

Is it possible to move this field to eMoF to do what Daphnis suggested in [ #10 (comment) 1](https://github.com/iobis/env-data/issues/10#issuecomment-347899317)?

In the meantime we can work with [ #10 (comment) 2](https://github.com/iobis/env-data/issues/10#issuecomment-354739951) although this only works if there is only one measurement per specimen. Assigning different occurrenceIDs to each specimen wouldn't work for this last case either

**Comment by pieterprovoost** [comment]
@rubenpp7 Can you please remind me why we cannot use proposal 2 here: https://github.com/iobis/env-data/issues/10#issuecomment-444422701 (in case of multiple measurements per organism, create separate occurrence records)?

I would avoid adding more fields to the extension as much as possible, as this takes us farther away from the core of the standard.

**Comment by rubenpp7** [comment]
@pieterprovoost If we assign different occurrenceIDs to each specimen we could store specimen data such as length and width, but how can we store "grouped" data of several specimens such as biomass or density? 

I am about to receive a dataset with length measurements of each individual but also an overall density per species.
To which occurrenceID do we link the densities?

**Comment by pieterprovoost** [comment]
Something like this?

Event:

| eventID | eventDate  | decimalLongitude | decimalLatitude |
| ------- | ---------- | ---------------- | --------------- |
| a       | 2020-01-01 | 3                | 52              |

Occurrence:

| eventID | occurrenceID | organismID | scientificName |
| ------- | ------------ | ---------- | -------------- |
| a       | 1            |            | Abra alba      |
| a       | 2            | x          | Abra alba      |
| a       | 3            | y          | Abra alba      |
| a       | 4            | z          | Abra alba      |

MeasurementOrFact:

| eventID | occurrenceID | measurementType | measurementValue | measurementUnit |
| ------- | ------------ | --------------- | ---------------- | --------------- |
| a       | 1            | biomass         | 5                | gC m-2          |
| a       | 1            | density         | 1000             | individuals m-2 |
| a       | 2            | length          | 10               | mm              |
| a       | 2            | width           | 8                | mm              |
| a       | 3            | length          | 12               | mm              |
| a       | 3            | width           | 9                | mm              |
| a       | 4            | length          | 13               | mm              |
| a       | 4            | width           | 10               | mm              |

Optionally you could add `individualCount` in Occurrence or `count` measurements in MeasurementOrFact.

**Comment by timvdstap** [comment]
Hi all, I'm interested in reviving this issue, as we had an interesting discussion earlier today. 

The issue we're running into is how best to structure the data when the occurrences are recorded different levels and nested. As an example, often times during trawl surveys, species' total weight and total catch is recorded, and then a subset (which, depending on the number of individuals caught might include all the species' individuals) is taken for individual measurement on length and weight. I'm mainly worried about the possibility of duplicating counts in OBIS or GBIF. It would seem that one occurrenceID cannot link to both the overall _and_ individual measurements (see table A below), as it would duplicate count information? But by creating a unique occurrenceID for individuals we may lose the information that it's a subset of the larger catch data info.  

eventID | occurrenceID | measurementType | measurementValue | measurementUnit
----------|  --------------| ------------------- | ---------------------- | ---------------------- 
a | 1 | Total abundance | 50 | individuals
a | 1 | Total weight | 10 | kg
a | 1 | individualCount | 1 | individuals
a | 1 | length | 60 | cm
a | 1 | weight | 500 | gr

Two solutions put forward were to e.g. 'hyphenate' the occurrenceID (sorry I don't know the proper term), so readers might infer that it's a subset. So that could look like: 

eventID | occurrenceID | measurementType | measurementValue | measurementUnit
----------|  --------------| ------------------- | ---------------------- | ---------------------- 
a | 1 | Total abundance | 50 | individuals
a | 1 | Total weight | 10 | kg
a | 1-1 | individualCount | 1 | individuals
a | 1-1 | length | 60 | cm
a | 1-1 | weight | 500 | gr

Or, alternatively, a distinction is made at the event level, where a 'subsample' is included as an additional layer:

eventID | occurrenceID | measurementType | measurementValue | measurementUnit
----------|  --------------| ------------------- | ---------------------- | ---------------------- 
a | 1 | Total abundance | 50 | individuals
a | 1 | Total weight | 10 | kg
a-1 | 2 | individualCount | 1 | individuals
a-1 | 2 | length | 60 | cm
a-1 | 2 | weight | 500 | gr

How can we best deal with these types of data, whereby we don't duplicate the counts of individuals whilst at the same time making sure that users can infer that some data is a subset of another occurrenceID? I hope I make sense and that I didn't forget to include any information! 

Thanks in advance!

**Comment by 7yl4r** [comment]
Looks like an interesting problem with a lot of good discussion.

Sub-occurrences are bound to cause confusion no matter how they get added. I would strongly advocate for an added column (`parentOccrrenceID`?) instead of a hyphen in `occurenceID`.

Something I've been thinking about is how analyses of a set of occurrences might be represented and shared. There are datasets with multiple occurrence records that could be summarized with a biomass calculation but do not have it. For example I submitted a zooplankton dataset as occurrence only, each species observation being a separate record. Biomass and abundance was not calculated, but it could be. These calculations could even be done by a third party. But how could a third party submit that?

A table could be added specifically for the purpose of describing a summary of occurrence records. This might be outside of OBIS's focus, I don't think there is anything like it in Darwin Core, and it's going to be a mess having an array column for `occurrenceRecordsSummarized`. But that is the approach that makes most sense to me. Maybe for OBIS that means preferring the length measurements on each occurrence and leaving the summarizing abundance and biomass implied.

**Comment by wardappeltans** [comment]
We do not recommend adding more terms to the eMoF extension. I think the last option by @timvdstap is our recommendation. You can add event hierarchy in the Event Core if you want to group occurrences and measurements.

**Comment by timvdstap** [comment]
@wardappeltans Thank you Ward! We will also have to make sure that the metadata sufficiently describes the subsetting of data.

**Comment by AbbyBenson** [comment]
@tucotuco I'm wondering if this issue might help with identifying needs for the new data model. I'm adding it to the narrative you requested but wanted to put it on your radar this way as well.

**Comment by tucotuco** [comment]
Thanks @albenson-usgs . It's already in there. :-)

**Comment by davidsean** [comment]
I came here because I have a similar situation. I also ended coming up with something similar to the last option by @timvdstap , as it is a **natural extension of the event hierarchy**. My samplings are broken down into sub sampling events, and the "ultimate" sub-sample comes down to picking out a single individual specimen. Here is a simplified breakdown (in my case I have at least two other levels of subsampling which I won't put here)

Notice that I have included `individualCount` as well as `organismID` in the Occurrence table to really make it clear that the target are individual specimens (not sure if I would do it this way for a published archive).

Event:
|eventID     |parentEventID |eventRemarks           |...|
|------------|--------------|-----------------------|---|
|a           |              |Trawl                  |   |
|a-sample    |a             |Trawl subsample        |   |
|a-1         |a             |single specimen sample |   |
|a-2         |a             |single specimen sample |   |
|a-3         |a             |single specimen sample |   |

Occurrence:

|eventID |occurrenceID   |individualCount    |organismID |scientificName |...|
|--------|---------------|-------------------|-----------|---------------|---|
|a       |a_1            |                   |           |Abra alba      |   |
|a-sample|a-sample-1     |10                 |           |Abra alba      |   |
|a-1     |a-1_1          |1                  |a-1        |Abra alba      |   |
|a-2     |a-2_1          |1                  |a-2        |Abra alba      |   |
|a-3     |a-3_1          |1                  |a-3        |Abra alba      |   |

eMoF:
|eventID |occurrenceID   |measurementType    |measurementValue   |measurementUnit    |...|
|--------|---------------|-------------------|-------------------|-------------------|---|
|a       |a_1            |biomass            |500                |kg                 |   |
|a-sample|a-sample_1     |biomass            |5                  |kg                 |   |
|a-1     |a-1_1          |length             |10                 |mm                 |   |
|a-1     |a-1_1          |width              |14                 |mm                 |   |
|a-1     |a-2_1          |length             |9                  |mm                 |   |
|a-1     |a-2_1          |width              |12                 |mm                 |   |
|a-1     |a-3_1          |length             |13                 |mm                 |   |
|a-1     |a-3_1          |width              |10                 |mm                 |   |

TLDR: The Event hierarchy can handle arbitrary subsampling levels. Picking out a single specimen can be seen as a type of subsampling. There is no need to add any supplemental fields.

EDIT: added parentEventID column

**Comment by tucotuco** [comment]
@davidsean That looks great. OBIS Extended Meaurements or Facts to the rescue! The only comment I would add is that dwc:organismQuantity and dwc:organismQuantityType are also available for the count information. There was an attempt once to deprecate dwc:individualCount as redundant once the other two terms came into existence, but that attempt could not reach consensus as dwc:individualCount was already being misused in another way in practice.

**Comment by timvdstap** [comment]
@davidsean @tucotuco Thank you for reviving this topic again, I like the approach where organismID is included in the occurrence extension. To further highlight the relationship between a subsample and the 'larger' sample, I have also been using the resourceRelationship extension -- would this (also) be considered a reasonable approach?

**Comment by tucotuco** [comment]
@timvdstap I have been looking above to find a ResourceRelationship approach. But I don't see one. What are you relating, and how?

**Comment by davidsean** [comment]
Perhaps a `relationshipOfResource` term like "sampled from" (I am not aware of a controlled term, but perhaps there is one) could further specify that specimen of Occurrence `a-1_1` is taken from Occurrence `a_1`?

This information can be infered in the event parent-child relationship---oh no! I just noticed I forgot to put parentEventID in my example.

**Comment by timvdstap** [comment]
@tucotuco See for example [this record in OBIS](https://ipt.iobis.org/obiscanada/resource?r=trawl-catch-and-specimen-data-from-the-shimada-during-the-2022-international-year-of-the-salmon-expedition), where I related the subsample (individual fish) to the overall catch. This relationship is described by the relationshipOfResource (`is a subset of`), connecting the two occurrenceIDs (which are represented by 'resourceID' and 'relatedResourceID'. 

edit: @davidsean Hmm.. I didn't think there was a controlled vocabulary for that field (or perhaps they have since changed that), but apparently there is (see [here](https://dwc.tdwg.org/terms/#relationshipOfResource)), so I might have to update that record accordingly to e.g. `located within`.

**Comment by tucotuco** [comment]
There is a lot unfolding here. There are a lot of potential solutions, but I want to avoid saying too much in an effort to avoid creating unnecessary confusion. I guess the good news is that there is no particular obstacle to expressing length measurements and abundances at multiple levels of material samples in a fairly simple way when using a Darwin Core Archive with an Event Core, Occurrence extension, and Extended Measurements or Facts Extension, taking advantage of the Event hierarchical capability. 

@timvdstap Thank you for that reference. I see what you are doing there. I am not a fan of the ResourceRelationship Extension, especially if it isn't needed. In your case, I don't think it is. The reason I think so is that the reference dataset is constructed without taking advantage of the [latest version of Darwin Core Event Core](https://rs.gbif.org/core/dwc_event_2024-02-19.xml). Specifically, the Events don't take advantage of parentEventIDs or eventTypes. 

@davidsean It appears that you are not using the eventType either, but could take good advantage of it instead of using eventRemarks for that purpose, something like this:
 
| eventID | parentEventID | eventType |
| --------- | ------------- | ----------|
| a | | Trawl |
| a-sample | a | Trawl subsample |
| a-1 | a | single specimen sample |
| a-2 | a | single specimen sample |
| a-3 | a | single specimen sample |

I don't know if it was intended that the single specimen samples should be direct children of the Trawl. They might otherwise have been direct children of the Trawl subsample by using "a-sample" for their parentEventIDs, but that really depends on what really happened "in the field".

The parent-child relationship expressed explicitly in the parentEventID (and corresponding eventID on the parent) should be sufficient to obviate the need for a ResourceRelationship record by virtue of the guiding principles of an Event hierarchy, captured in the document [Properties of hierarchical events in the Humboldt Extension for Ecological Inventories](https://eco.tdwg.org/hierarchy/). These principles apply well even without the Humboldt Extension.

Note that GBIF is trying to facilitate development of the [eventType vocabulary](https://registry.gbif-uat.org/vocabulary/EventType). Note also that vocabulary is not in production, and that it doesn't describe @davidseans case very well. The eventType vocabulary will be all important in the GBIF Unified Model, so there will be ample opportunity to participate in its development.

I'd like to make some observations about the Occurrence table, rendered thus:

| eventID | occurrenceID | individualCount | organismID | scientificName |
| ------- | ----------- | -------- | -------- | --------- |
| a | a_1 |   |   | Abra alba |
| a-sample | a-sample-1 | 10 |   | Abra alba |
| a-1 | a-1_1 | 1 | a-1 | Abra alba |
| a-2 | a-2_1 | 1 | a-2 | Abra alba |
| a-3 | a-3_1 | 1 | a-3 | Abra alba |

There is one thing in this rendition that is a concern, and that is the repetition of "a-1", "a-2", and "a-3" as identifiers for Events and for Organisms. For clarity I would maybe change the identifiers to be more easily distinguishable ("e", "e-sample", "e-1", "e-2", "e-3", then "org-1", "org-2", and "org-3" for the Organisms. While at it I would also change the Occurrence identifiers ("occ-sample-1", "occ-1", "occ-2", and "occ3" or something like that). This might seem pedantic, but clarity is useful. At first I thought the "a-1" and the "a_1" were the same, and I thought that was very forward thinking because with our GBIF Unified Model work we are now convinced that an Occurrence is just a special kind of Event (especially now that MaterialEntity properties have been transferred to that class from the Occurrence class), and therefore current occurrenceIDs could be used in the future as eventIDs. Until that happens though, it is probably best to continue to conceptualize them as distinct classes.

> edit: @davidsean Hmm.. I didn't think there was a controlled vocabulary for that field (or perhaps they have since changed that), but apparently there is (see [here](https://dwc.tdwg.org/terms/#relationshipOfResource)), so I might have to update that record accordingly to e.g. `located within`.

@timvdstap Be aware that those are just non-normative examples to illustrate how the term might work. There is currently no formal controlled vocabulary for the term dwc:relationshipOfResource. In the GBIF Unified Model this will be one of the other crucial controlled vocabularies to develop.

Sorry to drag on, but hopefully some of this is useful.

**Comment by timvdstap** [comment]
@tucotuco Thank you so much for your elaborate explanation. I agree that in hindsight perhaps a resourceRelationship extension was not necessary, and that this 'relationship' could have been expressed through parentEventID and eventType (and perhaps organismID as well). I'm a little confused by past-me in that one of my suggestions was to create unique eventIDs for individual taxonomic occurrences but apparently not having followed that approach when actually standardizing/mobilizing the data to OBIS...

Beyond this, there are still a few uncertainties around the use of eventType that I'd like clarification on before I make changes to the record in OBIS, see [here](https://github.com/iobis/manual/issues/92).

**Comment by tucotuco** [comment]
@timvdstap I have added [this comment](https://github.com/iobis/manual/issues/92#issuecomment-2211589968) in the issue you referenced.

**Comment by timvdstap** [comment]
Thank you @tucotuco - very helpful! Following your advice I have structured the event core and occurrence extension as follows: 

eventID | parentEventID | eventType
----------| -----------------| ----------|
SOC | | project
SOC-COS | SOC | station
SOC-COS-2022-6-12 | SOC-COS | stationVisit
SOC-COS-2022-6-12-m1 | SOC-COS-2022-6-12 | single specimen sample

This makes the event Core quite large, but I think indeed improves downstream user understanding of the structure of the data.

eventID | occurrenceID | individualCount | organismID | scientificName
----------|  --------------| ------------------- | ---------------------- | ---------------------- 
SOC-COS-2022-6-12 | SOC-COS-2022-6-12-occ-2 | 41 |  | Metacarcinus magister (Dana, 1852)
SOC-COS-2022-6-12-m1 | SOC-COS-2022-6-12-m1-occ-10 | 1 |  SOC-COS-2022-6-12-org-10 | Metacarcinus magister (Dana, 1852)
SOC-COS-2022-6-12-m2 | SOC-COS-2022-6-12-m2-occ-11 | 1 | SOC-COS-2022-6-12-org-11 | Metacarcinus magister (Dana, 1852)

Admittedly, I'm having a bit of a hard time conceptualizing the difference between the `organismID` and the `occurrenceID` -- although I guess an organismID can have multiple associated occurrenceIDs if it is recaptured. The organismID key itself however does not contribute to improved understanding on the nested structure, correct?

Once I have the record published I will reference it here as well.

**Comment by tucotuco** [comment]
@timvdstap This looks great.  It is all set up to take advantage of the Humboldt Extension as well, if absence of detection and abundance are of interest to be reported.

The Organism is the biological entity that persists through time, so you have it right that it might be associated with multiple Occurrences, each of which might report the Organism's traits at a place and time. If it isn't an Organism that can be identified between records, there isn't really any reason to mint an organismID for the Occurrence record - it is "some" Organism, not "this particular" Organism that might be seen again.

We are now convinced from the GBIF Unified Model work that an Occurrence is a special type of Event. It shared all the characteristics of an Event, and in addition adds the characteristics of an Organisms at that place and time. I anticipate that we will be able to say formally in coming data publishing models that an occurrenceID from the Occurrence record can be used directly as the eventID in its corresponding Event record, but that isn't quite how things work right now, so I think you have the structure and relationships safely expressed in your tables for the current paradigm.

**Comment by timvdstap** [comment]
Thank you @tucotuco. Actually given your clarification on `organismID` I think it might not be relevant for my dataset - the individuals measured cannot have subsequent occurrence events. 

I need to look into the Humboldt Extention a bit more as I do report on occurrence absence, but that is currently don't through the `occurrenceStatus` field.
